source/common/upstream/cluster_manager_impl.cc

```diff
@@ -835,24 +835,35 @@ ClusterManagerImpl::loadCluster(const envoy::config::cluster::v3::Cluster& clust
   auto& lb = new_cluster_pair_or_error->second;
   Cluster& cluster_reference = *new_cluster;
 
+  const auto cluster_info = cluster_reference.info();
```

Comment:
Yeah, you are complete right. Ref is enough here. But the `info()` will return a copy of the shared pointer rather than a ref.

So, it make no sense to use `const auto&` here, because we always create a copy of shared pointer. In the previous implementation, these code obviously create lots of unnecessary copies of shared pointer, we reduced the copy to just once.

---

By the way, I think lots of interface design could be better. For example, in the scenario to return a shared pointer member, lots of interface will return `shared_ptr<T>` directly, but not all callers alyways need the ownership. using `const shared_ptr<T>&` would be better. If we have some free time, I think we can re-evaluate these interfaces and create a clear specifications to guide these code.